pkg/networkservice/chains/forwarder/server.go

```diff
@@ -0,0 +1,136 @@
+	beforeConnectServers, afterConnectServers []networkservice.NetworkServiceServer
+	connectClients                            []networkservice.NetworkServiceClient
+	connectOptions                            []connect.Option
+}
+
+// Option modifies server option value
+type Option func(o *serverOptions)
+
+// WithName sets name of the NetworkServiceServer
+func WithName(name string) Option {
+	return func(o *serverOptions) {
+		o.name = name
+	}
+}
+
+// WithAuthorizeServer sets authorization server chain element
+func WithAuthorizeServer(authorizeServer networkservice.NetworkServiceServer) Option {
+	if authorizeServer == nil {
+		panic("Authorize server cannot be nil")
+	}
+	return func(o *serverOptions) {
+		o.authorizeServer = authorizeServer
+	}
+}
+
+// WithBeforeConnectServers sets before connect server chain element
+func WithBeforeConnectServers(beforeConnectServers ...networkservice.NetworkServiceServer) Option {
+	return func(o *serverOptions) {
+		o.beforeConnectServers = beforeConnectServers
+	}
+}
+
+// WithAfterConnectServers sets after connect server chain element
+func WithAfterConnectServers(afterConnectServers ...networkservice.NetworkServiceServer) Option {
+	return func(o *serverOptions) {
+		o.afterConnectServers = afterConnectServers
+	}
+}
+
+// WithConnectClients sets connect client chain element
+func WithConnectClients(connectClients ...networkservice.NetworkServiceClient) Option {
+	return func(o *serverOptions) {
+		o.connectClients = connectClients
+	}
+}
+
+// WithConnectOptions sets connect Options for the server
+func WithConnectOptions(connectOptions ...connect.Option) Option {
+	return func(o *serverOptions) {
+		o.connectOptions = connectOptions
+	}
+}
+
+// NewServer returns a new interpose endpoint chain
+func NewServer(ctx context.Context, tokenGenerator token.GeneratorFunc, clientURL *url.URL, options ...Option) endpoint.Endpoint {
+	opts := &serverOptions{
+		name:            "forwarder-" + uuid.New().String(),
+		authorizeServer: authorize.NewServer(authorize.Any()),
+	}
+	for _, opt := range options {
+		opt(opts)
+	}
+
+	rv := new(forwarder)
+	rv.Endpoint = endpoint.NewServer(ctx, tokenGenerator,
+		endpoint.WithName(opts.name),
+		endpoint.WithAuthorizeServer(opts.authorizeServer),
+		endpoint.WithAdditionalFunctionality(
+			chain.NewNetworkServiceServer(opts.beforeConnectServers...),
+			clienturl.NewServer(clientURL),
+			heal.NewServer(ctx,
+				heal.WithOnHeal(addressof.NetworkServiceClient(adapters.NewServerToClient(rv))),
+				heal.WithOnRestore(heal.OnRestoreIgnore)),
+			connect.NewServer(ctx,
+				client.NewClientFactory(
+					client.WithName(opts.name),
+					client.WithAdditionalFunctionality(
+						mechanismtranslation.NewClient(),
+						chain.NewNetworkServiceClient(opts.connectClients...),
+					),
+				),
+				opts.connectOptions...,
+			),
+			chain.NewNetworkServiceServer(opts.afterConnectServers...),
+		),
+	)
+
+	return rv
+}
```

Comment:
Did you also consider having APIs to create grpc.Server and register endpoint with NS registry using default transport credentials from source ? this would also avoid boilerplate code across the forwarders.

Reply:
It sounds interesting, probably it would be better to discuss such thing with @denis-tingaikin or @edwarnicke.